storage/rocksdb/ha_rocksdb.cc

```diff
@@ -10402,32 +10450,62 @@ void ha_rocksdb::get_auto_increment(ulonglong off, ulonglong inc,
     off = 1;
   }
 
-  longlong new_val;
+  Field *field;
+  ulonglong new_val, max_val;
+  field = table->key_info[table->s->next_number_index].key_part[0].field;
+  max_val = rdb_get_int_col_max_value(field);
 
   // Local variable reference to simplify code below
-  std::atomic<longlong> &auto_incr = m_tbl_def->m_auto_incr_val;
+  auto &auto_incr = m_tbl_def->m_auto_incr_val;
 
-  if (inc == 1 && off == 1) {
+  if (inc == 1) {
+    DBUG_ASSERT(off == 1);
     // Optimization for the standard case where we are always simply
     // incrementing from the last position
 
     // Use CAS operation in a loop to make sure automically get the next auto
-    // increment value while ensuring tha we don't wrap around to a negative
+    // increment value while ensuring that we don't wrap around to a negative
     // number.
+    //
+    // We set auto_incr to the min of max_val and new_val + 1. This means that
+    // if we're at the maximum, we should be returning the same value for
+    // multiple rows, resulting in duplicate key errors (as expected).
+    //
+    // If we return values greater than the max, the SQL layer will "truncate"
+    // the value anyway, but it means that we store invalid values into
+    // auto_incr that will be visible in SHOW CREATE TABLE.
     new_val = auto_incr;
-    while (new_val != std::numeric_limits<longlong>::max()) {
-      if (auto_incr.compare_exchange_weak(new_val, new_val + 1)) {
+    while (new_val != std::numeric_limits<ulonglong>::max()) {
+      if (auto_incr.compare_exchange_weak(new_val,
+                                          std::min(new_val + 1, max_val))) {
         break;
       }
     }
   } else {
-    // The next value can be more complicated if either `inc` or 'off' is not 1
-    longlong last_val = auto_incr;
+    // The next value can be more complicated if either 'inc' or 'off' is not 1
+    ulonglong last_val = auto_incr;
 
     // Loop until we can correctly update the atomic value
     do {
-      if (((last_val - off) / inc) ==
-          (std::numeric_limits<longlong>::max() - off) / inc) {
+      // Calculate the next value in the auto increment series: offset
+      // + N * increment where N is 0, 1, 2, ...
+      //
+      // For further information please visit:
+      // http://dev.mysql.com/doc/refman/5.7/en/replication-options-master.html
+      //
+      // The following is confusing so here is an explanation:
+      // To get the next number in the sequence above you subtract out the
+      // offset, calculate the next sequence (N * increment) and then add the
+      // offset back in.
+      //
+      // The additions are rearranged to avoid overflow.
+      //
+      DBUG_ASSERT(last_val > 0);
+      // The following is equivalent to (last_val - 1 + inc - off) / inc.
+      ulonglong n =
+          (last_val - 1) / inc + ((last_val - 1) % inc + inc - off) / inc;
```

Comment:
I'm having some trouble reasoning through the correctness of this calculation. For example, suppose

last_val = 75
max_val = 128
inc = 3
offset = 120

I'd expect the next value to be 120, followed by 123, 126 and then we hit the max. Is this correct? The statement indicates we generate sequences of offset + N * inc, where N in [0, 1, 2, 3...], so 120, 123, 126, 129...

However: (75 - 1) / 3 + ((75 - 1) % 3 + 3 - 120) / 3 = -14

Later, we have new_val * inc + offset: -14 * 3 + 120 = 78, which seems different from the original code.

Reply:
There's an assumption that off <= inc, which would make the (inc - off) calculation always return a non-negative number. I can add a DBUG_ASSERT if needed, but the assumption is enforced a few lines above in the same function.

Here's the reasoning. I'm using the fact that `(a+b)/c = a/c + b/c + (a%c + b%c) / c`. Here's a short proof:

```
(a+b)/c
= (a - a%c + a%c + b - b%c + b%c) / c
= (a - a%c) / c + (b - b%c) / c + (a%c + b%c) / c
= a/c + b/c + (a%c + b%c) / c
```

Now let `a = last_val -1` and `b = inc - off` and `c = inc` to get my statement.

Reply:
Ah, okay. The check will skip the degenerate case, so the rest of the logic follows then.